hnn_core/network.py

```diff
@@ -448,6 +508,82 @@ def update_types(self, gid_dict):
             spike_types += [list(spike_types_trial)]
         self._types = spike_types
 
+    def update_trial_bounds(self, tstart, tstop):
+        """Update tstart and tstop in the current instance of Spikes.
+
+        Parameters
+        ----------
+        tstart : int | float | None
+            Value defining the start time of all trials.
+            Only relevant for legacy files.
+        tstop : int | float | None
+            Value defining the stop time of all trials.
+            Only relevant for legacy files.
+
+        """
+
+        # Validate tstart, tstop
+        if not isinstance(tstart, (int, float)) or not isinstance(
+                tstop, (int, float)):
+            raise ValueError('tstart and tstop must be of type int or float')
+        elif tstop <= tstart:
+            raise ValueError('tstop must be greater than tstart')
+        self._tstart = tstart
+        self._tstop = tstop
+
+    def mean_rates(self, mean_type='all'):
+        """Mean spike rates (Hz) by cell type.
+
+        Parameters
+        ----------
+        mean_type : str
+            'all' : Average over trials and cells
+                Returns mean rate for cell types
+            'trial' : Average over cell types
+                Returns trial mean rate for cell types
+            'cell' : Average over individual cells
+                Returns trial mean rate for individual cells
+
+        Returns
+        -------
+        spike_rate : dict
+            Dictionary with keys 'L5_pyramidal', 'L5_basket', etc.
+        """
+        cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+        spike_rates = dict()
+        all_spike_types = np.array(sum(self._types, []))
+        all_spike_gids = np.array(sum(self._gids, []))
+
+        if mean_type not in ['all', 'trial', 'cell']:
+            raise ValueError("Invalid mean_type. Valid arguments include "
+                             "'all', 'trial', or 'cell'.")
+
+        for cell_type in cell_types:
+            type_mask = np.in1d(all_spike_types, cell_type)
+            cell_type_gids = np.unique(all_spike_gids[type_mask])
```

Comment:
It looks like you only consider only the gids that spike somewhere in at least one trial rather than *all* gids of a given type. For example, one might find that only 2 L5 pyramidal cells fire out of the entire network of L5 pyramidal cells, across trials. Thus, I think your code here will report an average spike rate across cells as the average of the 2 cells that fired rather than the 100 L5 pyramidal cells in the network that receive an input. Is this correct?

Reply:
In order to rectify this, you'll probably need to pass in the `Network.gid_dict` attribute. I'm beginning to wonder if this function belongs in the hnn_core analysis module.

Reply:
Great catch Ryan! That is correct, the mean rates would be incorrectly calculated when averaging across trials.

Also I may have had the same thought at [one point](https://github.com/jonescompneurolab/hnn-core/pull/155#issuecomment-684958576) ;)

In either case I think using `Network.gid_dict` would greatly simplify the calculations. At the outset the function could make a matrix that contains rates for every possible gid and trial. Then averaging would just require filtering by the gid index.

Reply:
another alternative is to have `gid_dict` as part of the `Spikes` class, although I'm not sure how you would save it to files then ...

Reply:
If we move forward with placing attributes as comments at the top, it wouldn't be too unreasonable to add `gid_dict` directly since it's just a list of cell types and ranges. 

The two major decisions at the moment are: 
- [ ] Decide on file format for saving comments at the top (applies to the other `write` functions as well)
- [ ] Keep this function in the `Spikes` class or move it to an analysis module

I'm going to hold off on pushing more commits until that's settled. Let me know if you think both answers are pretty clear or if they warrant more discussion.

Reply:
thinking again, I think `gid_dict` should just be passed to the function ... it can be `None` by default but must be a `dict` if you want to average by cell type. Same with `tstart` and `tstop` which should be required arguments. So the user has to be a bit careful about keeping track of these. But that way we don't worry about saving these information from `Spikes` object and can hold off the file format discussion for later and can move on with this PR.

as for the function, keep it in here for now. But let's keep a running record of what needs to go into the `analysis` module. Maybe in #139 ? When we have enough functions, we can move stuff there.

Reply:
That sounds good to me. I'll start working on this and remove the modifications to the `write` function.

Reply:
Given @rythorpe's comment, I think you need an account of every gid for all calculations. The only exception would be making the default behavior return mean rates for each individual gid with `mean_types='cell'`.

The current commit requires a gid_dict in all cases.